docs/v3/tutorials/airflow.mdx

```diff
@@ -0,0 +1,502 @@
+- **Logging & Observability**: Use Prefect's **UI** for logs or configure **external storage (e.g., S3, Elasticsearch)** for long-term retention.
+- **Notifications & Alerts**: Set up **failure alerts** via **Slack, email, or Twilio** using **Prefect Automations** to ensure timely issue resolution.
+- **CI/CD & Testing**: Validate **deployment YAMLs** in CI (`prefect deployment build --skip-upload`), and unit test tasks as **regular Python functions**.
+- **Configuration Management**: Replace Airflow **Variables/Connections** with **Prefect Blocks**, storing secrets in **CLI, UI, or version-controlled JSON**.
+- **Security & Access Control**: Prefect Cloud includes **built-in authentication & role-based access**; self-hosted setups should secure API and workers accordingly.
+- **Decommissioning Airflow**: Once migration is complete, **disable DAGs, archive the code, and shut down Airflow components** to reduce operational overhead.
+
+For more details on operating Prefect in production, see the [Operate](/v3/manage/index) guide.
+
+## Testing & Validation
+Thorough testing ensures your Prefect flows perform like their Airflow equivalents. Since this is a **full migration**, validation is essential before decommissioning Airflow.
+
+**Testing Prefect Flows in Isolation**
+- **Unit test task logic** - Write tests for tasks as regular Python functions.
+- **Run flows locally** - Execute `@flow` functions directly (`etl_pipeline_flow()`) to validate logic.
+- **Use Prefect's local orchestration** - Start a Prefect server (`prefect server start`), register a deployment, and trigger flows via Prefect UI to mirror production behavior.
+- **Compare outputs** - Run both Airflow and Prefect for the same input and validate results (e.g., database rows, file outputs). Debug discrepancies early.
+
+**Validation Phase: Temporary Parallel Running (Shadow Mode)**
+- **Keep the Airflow DAG inactive** but available for testing.
+- **Manually trigger** both Airflow and Prefect flows for the same execution date.
+- **Write test outputs separately** to prevent conflicts, ensuring parity before stopping Airflow runs.
+
+For batch jobs, this phase should be **brief**, ensuring correctness without long-term dual maintenance.
+
+**Decommissioning Airflow**
+Once a Prefect flow is stable, **disable the corresponding Airflow DAG** to prevent accidental execution. Clearly document Prefect as the new source of truth. Avoid keeping inactive DAGs indefinitely, as they can cause confusion—**archive or remove them once the migration is complete**.
+
+### Common issues and troubleshooting
+
+- **Missing dependencies:** If a Prefect flow fails with `ImportError`, ensure all required libraries are installed in the execution environment (Docker image, VM, etc.), not just locally.
+- **Credentials & access:** Verify that Prefect tasks have the same permissions as Airflow (e.g., service accounts, IAM roles). If using Kubernetes, ensure pods can access necessary databases and APIs.
```

Comment:
nit - do tasks have permissions?  Perhaps better to say workers here?

Reply:
Good call. Fixed